python/tskit/trees.py

```diff
@@ -1411,23 +1411,32 @@ def num_tracked_samples(self, u=None):
         return sum(self._ll_tree.get_num_tracked_samples(root) for root in roots)
 
     def _preorder_traversal(self, u):
-        stack = [u]
-        while len(stack) > 0:
-            v = stack.pop()
-            if self.is_internal(v):
-                stack.extend(reversed(self.children(v)))
+        stack = collections.deque([u])
+        # For perf we store these to avoid lookups in the tight loop
+        pop = stack.pop
+        extend = stack.extend
+        get_children = self.children
+        while stack:
```

Comment:
Minor style quibble: we prefer to avoid 'Falsey', unless we're really using the fact that might evaluate to False in different contexts. Here, I find ``while len(stack) > 0`` to be much more readable: it's totally explicit, and the reader doesn't have to think "is this going to be False because stack is a boolean, or it might be None, or it might be a list of length zero?". It's pretty obvious here since the method is so short, but still I think it's easier to read when we're totally explicit about the conditional.

Reply:
Completely agree - the intent is much clearer with `while len(stack) > 0`. 

I've used `while stack` here as you take ~10% perf hit from using the former as it has an extra python function call (2 calls for `postorder` if you also replace the `if children`). 

This level of optimisation might be overkill, given (as you said), that as soon as users start doing anything this part won't be the dominant term. Maybe a comment with the intent might be a good solution? Happy either way.

Reply:
Aha, I see. I didn't think there'd be that much difference. You're right, the best thing to do here is add a comment saying we do this for perf.

Reply:
Comments added, rebased to master.